### ISSUE TITLE
[$$ BOUNTY] Add Qwen 1.5 (0.5B) model to TT-Buda Model Demos

### ISSUE THREAD

**Issue by Shubhamsaboo** [tenstorrent/tt-buda-demos#20]
### **Background:**
TT-Buda model demos, developed by Tenstorrent, is a growing collection of model demos showcasing the capabilities of AI models running on Tenstorrent hardware. These demonstrations cover a wide range of applications, aiming to provide insights and inspiration for developers and researchers interested in advanced AI implementations.

### **Bounty Objective:**
We are excited to announce a bounty for contributing a new AI model demonstration to the TT-Buda repository. This is an opportunity for AI enthusiasts, researchers, and developers to showcase their skills, contribute to cutting-edge AI research, and earn rewards.

### Task Details: 

Integrate [**Qwen-1.5 (0.5B)**](https://huggingface.co/Qwen/Qwen1.5-0.5B) into the TT-Buda model demonstrations.

#### **Requirements:**

- The submission must include a comprehensive README.md detailing the model's architecture, implementation details, and usage instructions.
- The model should be fully functional and tested on Tenstorrent hardware, ensuring compatibility and performance optimization.
- Include sample inputs and outputs, demonstrating the model's capabilities.
- Provide documentation on any dependencies and installation procedures.

#### **Contribution Guidelines:**

- Fork the [TT-Buda model demos repository](https://github.com/tenstorrent/tt-buda-demos/tree/main/model_demos).
- Create a new directory within the `model_demos` folder following the naming convention: `model_yourModelName`.
- Ensure your code adheres to the coding standards and guidelines provided in the repository's `CONTRIBUTING.md` file.
- Submit a pull request with a detailed description of your model and any relevant information that will help reviewers understand and evaluate your contribution.

#### **Evaluation Criteria:**

- Innovation and Relevance: How does the model contribute new ideas or solutions? Is it relevant to current challenges in AI?
- Implementation Quality: Code readability, structure, and adherence to best practices.
- Performance: Efficiency and performance on Tenstorrent hardware.
- Documentation: Clarity and completeness of the accompanying documentation.

#### **Rewards:**
Contributions will be evaluated by the Tenstorrent team, and the best contribution will be eligible for $500 bounty. 

#### Get Started with Grayskull DevKit
Dive into AI development with the Grayskull DevKit, your gateway to exploring Tenstorrent's hardware. Paired with TT-Buda and TT-Metalium software approaches, it offers a solid foundation for AI experimentation. Secure your kit [here](https://tenstorrent.com/cards/?utm_source=bounty&utm_medium=github&utm_campaign=GS).

#### Connect on Discord
Join our [Discord](https://discord.gg/tenstorrent) to talk AI, share your journey, and get support from the Tenstorrent community and team. **Let's innovate together!**

**Comment by JushBJJ** [comment]
@Shubhamsaboo I want to have a crack at this but I don't have Grayskull, can I still do it?

**Comment by JushBJJ** [comment]
Update: I am working with @JonathanALevine to get this done for Qwen since he has a grayskull we can test it on

**Comment by Lewis300** [comment]
Hey all. So I do not have a Grayskull but I've been able to get the model to compile with decent results running with `PYBUDA_DEVMODE=1`.

Pytorch can use seemingly any tensor as a mask for `masked_fill` it just converts 0 to False and all other numbers to True. The Qwen implementation seems to use a mask in this manner at some call to `masked_fill`. 

So the big problem is that Qwen uses the lowest finite value a 32-bit float can represent as its mask filler. The way TVM decomposes `masked_fill` is by assuming that the condition mask is a boolean. However, for this model, this is not the case. This becomes problematic because as part of the decomposition of `masked_fill` in TVM we multiply the mask with the values of the input, and also the replacement value. At the end of the decompsition we will end up adding together `inf` and `-inf`, which creates a nan. And then in the following softmax returns all nans. 

So I've solved this by editing the decomposition of `masked_fill` so it may handle the event when the `mask` is not a boolean/one-hot tensor. This function already resides in `third_party/tvm/python/tvm/relay/frontend/pytorch.py::PyTorchOpConverter`. So you can just paste the body.

```
def masked_fill(self, inputs, input_types):
        mask = inputs[1]
        def replace_inf(inp, replacement_val=1e4):
            if isinstance(inp, _expr.Call) and list(_analysis.free_vars(inp)) == []:
                inp = _infer_value(inp, {}).asnumpy()
                inp[np.isneginf(inp)] = -replacement_val
                inp[np.isposinf(inp)] = replacement_val
                inp = _expr.const(tvm.nd.array(inp))
            elif isinstance(inp, float) and np.isinf(inp):
                inp = np.sign(inp)*replacement_val if np.isinf(inp) else inp
            return inp

        # Handle constant inputs where we can determine -inf/inf values from it
        inputs[0] = replace_inf(inputs[0])
        inputs[2] = replace_inf(inputs[2])

        if isinstance(mask, _expr.Call) and list(_analysis.free_vars(mask)) == []:
            mask = _expr.const(_infer_value(mask, {}))
        mask = _op.cast(mask, "float32")

        # Pybuda will convert all cast ops to Identity. This can be an issue.
        # In pytorch, when the mask has float values, it treats 0 as False and 
        # everything else as True. We cannot assume that mask will contain only
        # zeroes and ones, as multiplication with the mask will yeild incorrect
        # results.
        mask = _op.abs(mask)
        mask = _op.clip(mask, 0, 1.0)

        # In the event some values actually lie between 0 and 1 we use greater
        # to ceil those all to 1.
        # NOTE: The cast here is just to stop TVM from complaining. It will be
        # replaced with the identity function further down the line
        mask = _op.cast(_op.greater(mask, _expr.const(0, "float32")), "float32")

        value = _op.cast(_wrap_const(inputs[2]), input_types[0])
        value = _op.broadcast_to_like(value, mask)

        one_const = _expr.const(1, dtype="float32")
        inverse_mask = _op.subtract(one_const, mask)
        
        # (!mask * x) + (mask*y)
        return _op.add(_op.multiply(inputs[0], inverse_mask), _op.multiply(value, mask))
```

I will admit that `masked_fill` ends up becoming more ops than it was before. In most cases, the values, and mask passed to `masked_fill` are static. Thus these ops should be constevaled away during the consteval phase of compilation.

I've also edited the decomposition of `tril` so that it _should_ be able to decompose `tril` ops with a dynamic `diagonal` argument. Keyword _should_. I have not thoroughly tested this:

```
def tril(self, inputs, input_types):
        x = inputs[0]
        x_shape = _infer_shape(x)
        diagonal = inputs[1]

        count = np.arange(np.prod(x_shape)).reshape(x_shape)
        comp = count.transpose(-1, -2)
        count = tvm.relay.Constant(tvm.nd.array(count))
        count = _op.cast(count, self.infer_type(diagonal).dtype)
        comp = tvm.relay.Constant(tvm.nd.array(comp))
        comp = _op.cast(comp, self.infer_type(diagonal).dtype)

        tril = _op.less_equal(comp, _op.add(count, diagonal))
        tril = _op.cast(tril, self.infer_type(x).dtype)
        return _op.multiply(tril, x)
```

The model fully compiles, and through net2pipe as I've set `PYBUDA_VERIFY_NET2PIPE=1` (only required since I do not actually have TT hardware). The results I've been able to obtain are:

```
Input: My name is Lewis, nice to meet you.

PT answer: My name is Lewis, nice to meet you. I am a 12-year-old boy. I am from the USA. I am in Class 3, Grade 7. I have a good friend. His name is Tom. He is from England. He is in Class 2,

TT answer: My name is Lewis, nice to meet you. I am a 12-year-old boy. I am from the USA. I am in Class 3, Grade 7. I have a good friend. His name is Tom. He is考点','');
 a a  a there there there you
```

I am unsure where the model goes wrong at the end there. I'm confident its a separate issue though.

Good luck to whoever picks this up!

**Comment by JushBJJ** [comment]
👀

**Comment by JonathanALevine** [comment]
@JushBJJ and I took a look at this and I think this is on the right track. Some comments on implementation.

* pybuda version: tt-buda-0.10.5.gs.240315
* transformers version: 4.38.0 (required to get qwen model definitions and configs)
* Used @JushBJJ `triu` patch from here: https://github.com/tenstorrent/tt-tvm/pull/2
* Used `mask_fill` and `tril` patches from @Lewis300.
* Env vars:
```
os.environ['PYBUDA_DEVMODE'] = '1'
os.environ['TT_BACKEND_TIMEOUT'] = '0'
os.environ["PYBUDA_FORK_JOIN_EXPAND_FORK_OUTPUT_BUF"] = "0"
os.environ["TT_BACKEND_OVERLAY_MAX_EXTRA_BLOB_SIZE"] = "65536"
```
* compiler configs:
```
  # Set PyBuda configurations
  compiler_cfg = pybuda.config._get_global_compiler_config()
  compiler_cfg.default_df_override = pybuda._C.DataFormat.Float16_b
  compiler_cfg.amp_level = 2
```

**Comment by JonathanALevine** [comment]
Ran this again setting `PYBUDA_DEVMODE=1` I get this output:
```
Prefix text: My name is Thomas and my main
Generated text:
dict_values(['My name is Thomas and my main goal is to help you find the best deals on your next vacation. I have been in the travel industry for over 10 years and have worked with many different types of travel agencies. My goal with this website is for you to find a travel agency that will help make your vacation a success. We will work with you and your family to make sure that you have a great vacation and that your travel experience is as enjoyable as possible.'])
```

**Comment by Lewis300** [comment]
Ah so the log messages you are seeing involving cpu_fallback are just for a few operations near the start and end of the model which cannot execute on the device and must be run on CPU. The vast majority of the heavy lifting is done on the TTDevice (grayskull). However, when you use PYBUDA_DEVMODE=1 it may actually run the TTDevicr code on your cpu as well, I’m unsure. It’s best to go without using that if you have the card.

If running the exact same code with PYBUDA_DEVMODE=1 is what gave you the better answer there then you might have to tinker with some other env variables, default data format, amp level, etc…

I wonder if the Float16_b conversion of the min float32 mask filler is causing problems…

**Comment by JonathanALevine** [comment]
Removed this: `compiler_cfg.default_df_override = pybuda._C.DataFormat.Float16_b`

Seems to have fixed the output. This is what I get:

```
Prefix text: My name is Thomas and my main
Generated text:
dict_values(['My name is Thomas and my main goal is to help you find the best deals on your next vacation. I have been in the travel industry for over 10 years and have worked with many different types of travel agencies. My goal with this website is for you to find a travel agency that will help make your vacation a success. We will work with you and your family to make sure that you have a great vacation and that your travel experience is as enjoyable as possible.'])
```

**Comment by JushBJJ** [comment]
Updating PR and adding new patches

**Comment by JushBJJ** [comment]
@Lewis300 thanks for that `PYBUDA_VERIFY_NET2PIPE=1` var, it's been incredibly hard for me to test out TVM patches without a card

**Comment by Lewis300** [comment]
@JonathanALevine You were able to get that output without PYBUDA_DEVMODE?

**Comment by JonathanALevine** [comment]
> @JonathanALevine You were able to get that output without PYBUDA_DEVMODE?

That's correct! 

```
os.environ['PYBUDA_DEVMODE'] = '0'
```

**Comment by Lewis300** [comment]
@JushBJJ @JonathanALevine Addressing your slowness concerns...

CPU fallback is necessary for performing the embeddings and the inverse of embeddings at the end of the model.
I believe most `transformers` models will output their hidden states as well as the final output. The compiler can't know that you don't care about the hidden states that are outputted and so will keep them around in memory to output at the end. You should add `output_hidden_states=False` to the model config when retrieveing it from `transformers`. This model also outputs some cached values, and intermediary attention scores too. So get rid of those when you retrieve the original pytorch model:

`model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen1.5-0.5B", return_dict=False, output_hidden_states=False, use_cache=False, output_attentions=False)`.

Hopefully this helps!

**Comment by Lewis300** [comment]
For me, I've noticed that moving the result from the TT section of the model, and the tail-end CPU fallback section takes some time. Moreover the tail-end CPU fallback is a matmul and a reshape - both of which could be run on the Grayskull. They get put on CPU because the weights of that matmul are the embedding weights. Which, I see how you'd want them on the same device for training purposes. For inference it seems to me that it would be alright to just clone them onto the device. 

Go to `third_party/tvm/python/tvm/relay/op/contrib/buda/buda.py` line 1273. And replace 
`fallback_nodes = add_shared_weights_to_fallback(graph_constructor.graph, graph_constructor.fallback_nodes, input_names)`
with
`fallback_nodes = graph_constructor.fallback_nodes`

This ends up removing the tail-end CPU fallback entirely for this model and with that, the delay. Maybe adding shared weight operations can be added to the compiler configuration. I guess that would be for TT to figure out.

**Comment by JonathanALevine** [comment]
To summarize a bit what was discovered today:

(1) `pybuda/op/eval/pybuda/tm.py` patch
@Lewis300 and I thought that the slow queue and dequeue between the TT part of the model and the tail-end CPU fallback part could be due to the precision of the model weights.  To do this with `model.to(torch.bfloat16) had to patch `tm.py` like this:
```
def eval(type, attr, ops):
    assert len(ops) == 1 or (type == "adv_index" and len(ops) == 2), f"Tensor manipulation ops should have one input {len(ops)} {attr}"
    t_ops = to_torch_operands(*ops)
    dtype = ops[0].dtype

    if type == "transpose":
        assert len(attr) == 3, "Transpose should have 3 attributes"
        dim0, dim1, orig_size = attr
        return torch.transpose(t_ops[0], dim0, dim1)

    if type == "reshape":
        return t_ops[0].reshape(attr)

    if type == "select":
        assert len(attr) == 4, "Select should have 4 attributes"
        dim, begin, length, stride = attr
        zero_shape = list(t_ops[0].shape)
        zero_shape[dim] = 1
        zero_slice = torch.zeros(zero_shape, dtype=dtype).squeeze(dim)
        result = []
        for offset in range(0, t_ops[0].shape[dim] - begin, stride):
            for i in range(begin, begin + length):
                if offset + i < t_ops[0].shape[dim] or stride == t_ops[0].shape[dim]:
                    result.append(t_ops[0].select(dim, offset + i))
                else:
                    result.append(zero_slice)
        return torch.stack(result, dim=dim)

    if type == "gather":
        assert len(attr) == 5, "Gather should have 5 attributes"
        dim, begin, length, stride, orig_size = attr
        x = t_ops[0]
        result = []
        zero_shape = list(x.shape)
        if dim > 0:
            dim -= 4
        while len(zero_shape) <= abs(dim):
            zero_shape = [1] + zero_shape
            x = x.unsqueeze(0)
        zero_shape[dim] = 1
        zero_slice = torch.zeros(zero_shape, dtype=dtype).squeeze(dim)
        offset = 0
        for i in range(0, orig_size):
            range_i = (i - begin) % stride
            if i >= begin and range_i < length:
                result.append(x.select(dim, offset))
                offset += 1
            else:
                result.append(zero_slice)
        return torch.stack(result, dim=dim)

    if type == "index":
        assert len(attr) == 4, "Index should have 4 attributes"
        dim, start, stop, stride = attr
        if dim >= 0:
            dim -= len(ops[0].shape)

        if dim == -5:
            return t_ops[0][..., start:stop:stride, :, :, :, :]
        elif dim == -4:
            return t_ops[0][..., start:stop:stride, :, :, :]
        elif dim == -3:
            return t_ops[0][..., start:stop:stride, :, :]
        elif dim == -2:
            return t_ops[0][..., start:stop:stride, :]
        elif dim == -1:
            return t_ops[0][..., start:stop:stride]
        else:
            raise NotImplementedError(f"Dim={dim}")

    if type == "adv_index":
        assert len(attr) == 1, "AdvIndex should have 1 attributes"
        dim = attr[0]
        assert dim == 0, "Currently not supported"

        if len(t_ops[1].shape) > 1:
            if len(t_ops[0].shape) > len(t_ops[1].shape) and t_ops[0].shape[0] == 1:
                # Padded
                # ret = torch.unsqueeze(t_ops[0][0][t_ops[1].numpy()], 0)
                ret = t_ops[0][0][t_ops[1].int()]
            else:
                # ret = torch.unsqueeze(t_ops[0][t_ops[1].numpy()], 0)
                ret = t_ops[0][t_ops[1].int()]
        else:
            # ret = t_ops[0][t_ops[1].numpy()]
            ret = t_ops[0][t_ops[1].int()]
        return ret
```

**Comment by Shubhamsaboo** [comment]
@JushBJJ can you link your new PR for adding Qwen 0.5 to this issue.

**Comment by JushBJJ** [comment]
PR for Qwen: https://github.com/tenstorrent/tt-buda-demos/pull/37

**Comment by Shubhamsaboo** [comment]
Claimed by @JushBJJ. It will be closed once it's merged into main.

Congrats Jush!

**Comment by Shubhamsaboo** [comment]
Closing this one as merged to main. Congrats again @JushBJJ.
